File: voice-relay/voice_relay/oauth/store.py
"""State for the authorization server.

  - clients      DCR registrations (persisted — a connector shouldn't have
                 to re-register after a relay restart)
  - access/refresh tokens (persisted — a connector shouldn't be logged out
                 by a restart)
  - pending      in-flight /authorize requests keyed by our Google `state`
                 (in-memory, short-lived)
  - codes        our issued authorization codes (in-memory, short-lived,
                 one-time-use)

Persistence is a single JSON file (atomic write, 0600) when a path is
given, else pure in-memory (tests). Codes/pending are never persisted —
they live seconds.
"""

from __future__ import annotations

import json
import os
import tempfile
import threading
import time
from dataclasses import asdict, dataclass, field
from typing import Optional


@dataclass
class TokenRecord:
    slack_uid: str
    email: str
    client_id: str
    scopes: list
    expires_at: int          # epoch seconds; 0 = no expiry (refresh tokens)
    refresh_token: str = ""  # paired refresh token, for access tokens


class Store:
    def __init__(self, path: Optional[str] = None):
        self._path = path
        self._lock = threading.Lock()
        self._clients: dict = {}                  # client_id -> client JSON (dict)
        self._access: dict[str, TokenRecord] = {}  # token -> record
        self._refresh: dict[str, TokenRecord] = {}  # token -> record
        self._pending: dict[str, dict] = {}       # state -> pending (in-mem)
        self._codes: dict = {}                    # code -> HoneybotAuthorizationCode (in-mem)
        self._load()

    # ---- clients (persisted) ----------------------------------------------
    def get_client(self, client_id: str) -> Optional[dict]:
        with self._lock:
            return self._clients.get(client_id)

    def put_client(self, client_id: str, client_json: dict) -> None:
        with self._lock:
            self._clients[client_id] = client_json
            self._persist()

    # ---- pending authorize (in-memory) ------------------------------------
    def put_pending(self, state: str, data: dict) -> None:
        with self._lock:
            self._pending[state] = {**data, "created_at": time.time()}

    def pop_pending(self, state: str, max_age: float = 600) -> Optional[dict]:
        with self._lock:
            data = self._pending.pop(state, None)
        if not data:
            return None
        if time.time() - data.get("created_at", 0) > max_age:
            return None
        return data

    # ---- authorization codes (in-memory, one-time) ------------------------
    def put_code(self, code: str, obj) -> None:
        with self._lock:
            self._codes[code] = obj

    def get_code(self, code: str):
        with self._lock:
            return self._codes.get(code)

    def pop_code(self, code: str):
        with self._lock:
            return self._codes.pop(code, None)

    # ---- tokens (persisted) -----------------------------------------------
    def put_access(self, token: str, rec: TokenRecord) -> None:
        with self._lock:
            self._access[token] = rec
            self._persist()

    def get_access(self, token: str) -> Optional[TokenRecord]:
        with self._lock:
            return self._access.get(token)

    def put_refresh(self, token: str, rec: TokenRecord) -> None:
        with self._lock:
            self._refresh[token] = rec
            self._persist()

    def get_refresh(self, token: str) -> Optional[TokenRecord]:
        with self._lock:
            return self._refresh.get(token)

    def revoke(self, token: str) -> None:
        with self._lock:
            self._access.pop(token, None)
            self._refresh.pop(token, None)
            self._persist()

    # ---- persistence ------------------------------------------------------
    def _load(self) -> None:
        if not self._path or not os.path.exists(self._path):
            return
        try:
            with open(self._path, encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, json.JSONDecodeError):
            return
        self._clients = data.get("clients", {})
        self._access = {t: TokenRecord(**r) for t, r in data.get("access", {}).items()}
        self._refresh = {t: TokenRecord(**r) for t, r in data.get("refresh", {}).items()}

    def _persist(self) -> None:
        if not self._path:
            return
        payload = {
            "clients": self._clients,
            "access": {t: asdict(r) for t, r in self._access.items()},
            "refresh": {t: asdict(r) for t, r in self._refresh.items()},
        }
        try:
            os.makedirs(os.path.dirname(self._path), exist_ok=True)
            fd, tmp = tempfile.mkstemp(dir=os.path.dirname(self._path))
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                json.dump(payload, f)
            os.chmod(tmp, 0o600)
            os.replace(tmp, self._path)
        except OSError:
            pass

File: voice-relay/src/oauth/store.ts
/**
 * Authorization-server state.
 *   clients        DCR registrations (persisted — survive relay restart)
 *   access/refresh tokens (persisted — a connector isn't logged out by a restart)
 *   pending        in-flight /authorize keyed by our Google `state` (in-mem, short)
 *   codes          our issued auth codes (in-mem, short, one-time)
 * Persistence is a single JSON file (atomic, 0600) when a path is given.
 */

import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { randomBytes } from "node:crypto";
import type { OAuthClientInformationFull } from "@modelcontextprotocol/sdk/shared/auth.js";

export interface TokenRecord {
  slackUid: string;
  email: string;
  clientId: string;
  scopes: string[];
  expiresAt: number; // epoch ms
  refreshToken?: string;
}

export interface PendingAuth {
  clientId: string;
  redirectUri: string;
  codeChallenge: string;
  clientState?: string;
  scopes: string[];
  createdAt: number;
}

export interface StoredCode {
  code: string;
  clientId: string;
  redirectUri: string;
  codeChallenge: string;
  scopes: string[];
  slackUid: string;
  email: string;
  expiresAt: number;
}

export class OAuthStore {
  private clients = new Map<string, OAuthClientInformationFull>();
  private access = new Map<string, TokenRecord>();
  private refresh = new Map<string, TokenRecord>();
  private pending = new Map<string, PendingAuth>();
  private codes = new Map<string, StoredCode>();

  constructor(private readonly path?: string) {
    this.load();
  }

  // ---- clients (persisted) ----
  getClient(clientId: string): OAuthClientInformationFull | undefined {
    return this.clients.get(clientId);
  }
  putClient(client: OAuthClientInformationFull): void {
    this.clients.set(client.client_id, client);
    this.persist();
  }

  // ---- pending authorize (in-memory) ----
  putPending(state: string, data: Omit<PendingAuth, "createdAt">): void {
    this.pending.set(state, { ...data, createdAt: Date.now() });
  }
  popPending(state: string, maxAgeMs = 600_000): PendingAuth | undefined {
    const data = this.pending.get(state);
    this.pending.delete(state);
    if (!data || Date.now() - data.createdAt > maxAgeMs) return undefined;
    return data;
  }

  // ---- authorization codes (in-memory, one-time) ----
  putCode(code: string, obj: StoredCode): void {
    this.codes.set(code, obj);
  }
  getCode(code: string): StoredCode | undefined {
    return this.codes.get(code);
  }
  popCode(code: string): void {
    this.codes.delete(code);
  }

  // ---- tokens (persisted) ----
  putAccess(token: string, rec: TokenRecord): void {
    this.access.set(token, rec);
    this.persist();
  }
  getAccess(token: string): TokenRecord | undefined {
    return this.access.get(token);
  }
  putRefresh(token: string, rec: TokenRecord): void {
    this.refresh.set(token, rec);
    this.persist();
  }
  getRefresh(token: string): TokenRecord | undefined {
    return this.refresh.get(token);
  }
  revoke(token: string): void {
    this.access.delete(token);
    this.refresh.delete(token);
    this.persist();
  }

  // ---- persistence ----
  private load(): void {
    if (!this.path || !existsSync(this.path)) return;
    try {
      const data = JSON.parse(readFileSync(this.path, "utf-8")) as {
        clients?: Record<string, OAuthClientInformationFull>;
        access?: Record<string, TokenRecord>;
        refresh?: Record<string, TokenRecord>;
      };
      for (const [id, c] of Object.entries(data.clients ?? {})) this.clients.set(id, c);
      for (const [t, r] of Object.entries(data.access ?? {})) this.access.set(t, r);
      for (const [t, r] of Object.entries(data.refresh ?? {})) this.refresh.set(t, r);
    } catch {
      /* start empty */
    }
  }

  private persist(): void {
    if (!this.path) return;
    try {
      mkdirSync(dirname(this.path), { recursive: true });
      const payload = {
        clients: Object.fromEntries(this.clients),
        access: Object.fromEntries(this.access),
        refresh: Object.fromEntries(this.refresh),
      };
      const tmp = `${this.path}.${randomBytes(6).toString("hex")}.tmp`;
      writeFileSync(tmp, JSON.stringify(payload), { mode: 0o600 });
      renameSync(tmp, this.path);
    } catch {
      /* best-effort; 1Password remains the durable client seed */
    }
  }
}
